src/Server.cpp

```diff
@@ -925,6 +1106,115 @@ bool isCompleteBody(Connection& connection)
+		LOG_ERROR << "Error: write(): can't send to CGI: " + std::string(std::strerror(errno));
+		connection.m_request.httpStatus = StatusInternalServerError;
+		connection.m_status = Connection::BuildResponse;
+		server.modifyEvent(connection.m_clientFd, EPOLLOUT);
+		server.removeCGIFileDescriptor(server, connection.m_pipeFromCGIReadEnd);
+		server.removeCGIFileDescriptor(server, connection.m_pipeToCGIWriteEnd);
+		return;
+	}
+	if (bytesSent == static_cast<long>(connection.m_request.body.size())) {
+		LOG_DEBUG << "CGI: Full body sent";
+		connection.m_status = Connection::ReceiveFromCGI;
+		connection.m_request.body.clear();
+		server.removeCGIFileDescriptor(server, connection.m_pipeToCGIWriteEnd);
+		return;
+	}
+	connection.m_request.body = connection.m_request.body.substr(bytesSent);
+}
+
+/**
+ * @brief Handles receiving data from a CGI script and processes it accordingly.
+ *
+ * This function reads data from the CGI script's output pipe and appends it to the request body.
+ * It also handles errors during the read operation and manages the connection state based on the
+ * data received. If the read operation indicates that the CGI script has finished sending data,
+ * the function will transition the connection to the response-building state and clean up the
+ * associated file descriptors.
+ *
+ * @param server Reference to the Server instance managing the connection.
+ * @param connection Reference to the Connection instance representing the client connection.
+ */
+void connectionReceiveFromCGI(Server& server, Connection& connection)
+{
+	char buffer[Server::s_cgiBodyBufferSize] = {};
+	long bytesRead = read(connection.m_pipeFromCGIReadEnd, buffer, sizeof(buffer));
+
+	if (bytesRead == -1) {
+		LOG_ERROR << "Error: read(): can't read from CGI: " + std::string(std::strerror(errno));
+		connection.m_request.httpStatus = StatusInternalServerError;
+		connection.m_status = Connection::BuildResponse;
+		server.modifyEvent(connection.m_clientFd, EPOLLOUT);
+		server.removeCGIFileDescriptor(server, connection.m_pipeFromCGIReadEnd);
+		if (connection.m_pipeToCGIWriteEnd != -1)
+			server.removeCGIFileDescriptor(server, connection.m_pipeToCGIWriteEnd);
+		return;
+	}
+	if (bytesRead == 0) {
+		LOG_DEBUG << "CGI: Full body received";
+		connection.m_status = Connection::BuildResponse;
+		server.modifyEvent(connection.m_clientFd, EPOLLOUT);
+		server.removeCGIFileDescriptor(server, connection.m_pipeFromCGIReadEnd);
+		if (connection.m_pipeToCGIWriteEnd != -1)
+			server.removeCGIFileDescriptor(server, connection.m_pipeToCGIWriteEnd);
+		int status = 0;
+		if (waitpid(connection.m_cgiPid, &status, 0) == -1) {
+			LOG_ERROR << "Error: waitpid(): " + std::string(std::strerror(errno));
+			connection.m_request.httpStatus = StatusInternalServerError;
+			return;
+		}
+		// Any child exit status unequal to 0 indicates unsuccessful completion of the process
+		// NOLINTNEXTLINE misinterpretation by HIC++ standard
+		if (WEXITSTATUS(status) != 0) {
```

Comment:
If we save the return of WEXITSTATUS() we only need NOLINTNEXTLINE once

We could also check with WIFSIGNALED and WIFEXITED what caused the child to terminate?

Reply:
Use WIFEXITED to determine "normal" exit of child, then check WEXITSTATUS.

Added check if child terminated due to signals using WIFSIGNALED and WTERMSIG.